### PR TITLE
Vector splats

### PR DESCRIPTION
**Pull request by kvark**
See https://github.com/gpuweb/gpuweb/pull/1611

This PR adds the `Expression::Splat` variant to the IR. This is useful, versus relying on `Expression::Compose`, since the splat is much more limited, and these expressions do not require heap allocated component lists.

This is fairly trivial to support in the backends.
As for the frontends, WGSL has a bit of (necessary) complexity to auto-detect splats and build the proper IR. SPV-in also has a small piece to detect when a splat is used.

### CODE REVIEW

File: src/front/wgsl/mod.rs [hunk 1]
```diff
@@ -328,12 +328,50 @@ impl<'a> ExpressionContext<'a, '_, '_> {
...
+            &mut Lexer<'a>,
+            ExpressionContext<'a, '_, '_>,
+        ) -> Result<Handle<crate::Expression>, Error<'a>>,
+    ) -> Result<Handle<crate::Expression>, Error<'a>> {
+        let mut left = parser(lexer, self.reborrow())?;
+        while let Some(op) = classifier(lexer.peek().0) {
+            let _ = lexer.next();
+            let mut right = parser(lexer, self.reborrow())?;
+            // insert splats, if needed by the non-'*' operations
+            if op != crate::BinaryOperator::Multiply {
```
Comment (jimblandy) [inline src/front/wgsl/mod.rs:353]: The reason `Multiply` is handled specially here is that it handles splats in the IR, correct? I wonder if it wouldn't ultimately make the IR simpler to make all splats explicit. But either way, that shouldn't be part of this PR.

**Review by jimblandy — APPROVED** [review]
Looks good to me! I had a few questions to check my understanding, and a minor suggestion.

File: src/front/glsl/constants.rs [hunk 2]
```diff
@@ -64,6 +66,28 @@ impl<'a> ConstantSolver<'a> {
...
+                        self.types
+                            .fetch_if(|t| t.inner == crate::TypeInner::Vector { size, kind, width })
+                    }
+                    ConstantInner::Composite { .. } => None,
+                };
+                Ok(self.constants.fetch_or_append(Constant {
+                    name: None,
+                    specialization: None,
+                    inner: ConstantInner::Composite {
+                        ty: ty.ok_or(ConstantSolvingError::SplatType)?,
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/front/glsl/constants.rs:86]: This is checking if the type is registered. If we just have an expression, its type is not guaranteed to be registered in the type arena. In this case, we are talking about the `Splat` expression. We don't want to panic, strictly speaking, because Naga does not need to panic at all.

File: src/back/spv/writer.rs [hunk 3]
```diff
@@ -1384,17 +1371,36 @@ impl Writer {
...
+                id
+            }
             crate::Expression::Compose {
                 ty: _,
                 ref components,
             } => {
-                //TODO: avoid allocation
-                let mut constituent_ids = Vec::with_capacity(components.len());
+                self.temp_chain.clear();
                 for &component in components {
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/spv/writer.rs:1392]: holy magic! Never used splice before

File: src/front/spv/mod.rs [hunk 4]
```diff
@@ -1184,9 +1184,17 @@ impl<I: Iterator<Item = u32>> Parser<I> {
...
+                    let ty = self.lookup_type.lookup(result_type_id)?.handle;
+                    let first = components[0];
+                    let expr = match type_arena[ty].inner {
+                        // this is an optimization to detect the splat
+                        crate::TypeInner::Vector { size, .. }
+                            if components.len() == size as usize
+                                && components[1..].iter().all(|&c| c == first) =>
+                        {
+                            crate::Expression::Splat { size, value: first }
+                        }
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/front/spv/mod.rs:1196]: This will be checked during either the `Typifier` logic, or later in the IR validator. Frontend is not the place to validate it anyway.

File: src/front/wgsl/mod.rs (hunk 1 above)
Comment (kvark) [reply src/front/wgsl/mod.rs:353]: The IR currently follows SPIR-V, where "*" has a special place, because it's always nicely accelerated (i.e. multiply-add instructions, matrix-vector multiplication, etc), and also is sounds mathematically (linear scale). So even if we change the "*" to not be special for splats, it's still going to be special.

File: src/back/spv/writer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/spv/writer.rs:1392]: I tried it, and I haven't reached a solid conclusion on whether it's good to use here. First, because it's fairly magical: there is a range argument, a closure, and it returns an iterator. Its guarantees on performance depend on how it's used... Second, because `self.cached` can't be used inside the closure, so that's another line to spend, anyway.
Sometimes the dumb code is the best :)

[the remark this answers is not in the document]
Comment (jimblandy) [reply src/back/spv/writer.rs:1393]: Oh, that's too bad: the borrow checker doesn't see how to split the borrow, so the closure borrows all of `self`. This does work:
```
self.temp_list.clear();
let cached = &self.cached;
self.temp_list.extend(components.iter().map(|&c| cached[c]));
```
but at that point it's becoming arguable whether this is a clarification or not. Oh well.

[the remark this answers is not in the document]
Comment (kvark) [reply src/back/spv/writer.rs:1393]: yes, it's certainly workaroundable, just a bit less nice than originally desired :)
